### PR TITLE
[export] Implement the calling convention for exporting with multi-platform lowering

### PR DESCRIPTION
**Pull request by gnecula**
This is a first step towards supporting multi-platform exported JAX modules. Such modules are usable on more than one platform, and take an additional first argument that encodes the actual compilation platform as an index into the sequence of platforms for which the module was lowered. More details about the calling convention are in the docstring for jax_export.Exported in this PR.

The value of the platform index is set by `jax_export.call_exported` when calling from JAX, and in the tf.XlaCallModule prior to compilation, when called from TensorFlow. This is already implemented in tf.XlaCallModule.

This PR has some incomplete pieces:

  * Currently we actually lower only for the first platform specified, and the platform argument is not used. There are a couple of implementation strategies for actual multi-platform lowering, both using the same calling convention. We could lower separately for each platform and put the results together with one top-level conditional. Alternatively, we can take advantage of the fact that few primitives have per-platform lowering; we could lower those using a conditional.
  * we implement multi-platform lowering only for jax_export, not for regular JAX jit or AOT lowering. This ensure that this change is narrowly scoped and safe for most JAX usage.
  * we abuse the `_experimental_lowering_platform` kwarg to `lower()` to pass a tuple of platforms when we want multi-platform lowering. We ought to rename it to `_experimental_lowering_platforms`, but that requires more plumbing.
  * we take advantage of the fact that the lowering for the platform index is identical to that for dimension variables: add a new argument to inner functions and pass the values to callees. We implement platform index as a dimension variable.
  * we do not yet have the connection with jax2tf.convert.

### CODE REVIEW

**Comment by gnecula** [comment]
@junwhanahn @maxwillzq PTAL

File: jax/_src/interpreters/mlir.py [hunk 1]
```diff
@@ -472,7 +490,7 @@ def __init__(
                                         if cached_call_jaxpr_lowerings is None
                                         else cached_call_jaxpr_lowerings)
     self.override_lowering_rules = override_lowering_rules
-    self.shape_poly_state = shape_poly_state or ShapePolyLoweringState(())
+    self.shape_poly_state = shape_poly_state or ShapePolyLoweringState((), ('cpu',))
```
Comment (maxwillzq) [inline jax/_src/interpreters/mlir.py:493]: I have some questions about why the default value should be always "CPU" but not jax.default_backend()

File: jax/_src/interpreters/mlir.py [hunk 2]
```diff
@@ -646,7 +664,7 @@ def lower_jaxpr_to_module(
     jaxpr: core.ClosedJaxpr,
     ordered_effects: list[core.Effect],
     backend_or_name: str | xb.XlaBackend | None,
-    platform: str,
+    platform: str | tuple[str, ...],
```
Comment (maxwillzq) [inline jax/_src/interpreters/mlir.py:667]: nit:  If we change this argument name to "platforms", how many additional changes do we need to make ？ If not too many places, we should update it.

File: jax/experimental/jax2tf/jax_export.py [hunk 3]
```diff
@@ -154,6 +158,67 @@ class Exported:
...
+  Assume that we use multi-platform lowering, and we have
+  ordered effects. The `main` function will be as follows:
+
+      func public main(platform_index: i32, arg: f32[?, ?]) {
+         arg_w = hlo.get_dimension_size(arg, 0)
+         dim1 = hlo.get_dimension_size(arg, 1)
+         arg_h = hlo.floordiv(dim1, 2)
+         call _check_shape_assertions(arg)  # See below
+         token = new_token()
+         token_out, res = call _wrapped_jax_export_main(platform_index, arg_h, arg_w, token_in, arg)
```
Comment (maxwillzq) [inline jax/experimental/jax2tf/jax_export.py:190]: I did not get why the order `arg_w`, `arg_h` need change to 'arg_h', 'arg_w' here ? Any special reason or just typo ?
Thanks

File: jax/experimental/jax2tf/jax_export.py [hunk 4]
```diff
@@ -955,12 +990,14 @@ def _call_exported_impl(*args, exported: Exported):
 def _call_exported_lowering(ctx: mlir.LoweringRuleContext, *args,
                             platform: str,
                             exported: Exported):
-  if (platform != exported.lowering_platform and
+  # TODO: implement true multi-platform lowering for call_exported
+  if (platform not in exported.lowering_platforms and
       DisabledSafetyCheck.platform() not in exported.disabled_checks):
     raise ValueError(
         f"The exported function '{exported.fun_name}' was lowered for "
         f"platform '{exported.lowering_platform}' but it is used "
```
Comment (maxwillzq) [inline jax/experimental/jax2tf/jax_export.py:998]: nit: we also need update the error msg here "pltforms '{exported.lowering_platforms}'" to keep consistency.

**Review by maxwillzq — APPROVED** [review]

File: jax/_src/interpreters/mlir.py (hunk 2 above)
Comment (gnecula) [reply jax/_src/interpreters/mlir.py:667]: That is in the plan, but there are quite a few places where that is used, and I want to have as few changes as possible for now to the rest of JAX.

**Review by maxwillzq — APPROVED** [review]

**Review by junwhanahn — COMMENTED** [review]
One high-level question (not a blocker): what are the advantages of exporting one StableHLO module with multi-platform functions like in this PR, compared to simply exporting one StableHLO module per platform?

I'm wondering if we will ever be in a situation where having just one StableHLO module for all platforms could be limiting, e.g., different module-level params depending on platforms, having to do constant folding to get rid of irrelevant branches (especially if some custom calls are not known by other platforms), etc. If so, dispatching different StableHLO modules at the `XlaCallModule` level might be simpler/modular.

**Comment by gnecula** [comment]
> One high-level question (not a blocker): what are the advantages of exporting one StableHLO module with multi-platform functions like in this PR, compared to simply exporting one StableHLO module per platform?
> 
> I'm wondering if we will ever be in a situation where having just one StableHLO module for all platforms could be limiting, e.g., different module-level params depending on platforms, having to do constant folding to get rid of irrelevant branches (especially if some custom calls are not known by other platforms), etc. If so, dispatching different StableHLO modules at the `XlaCallModule` level might be simpler/modular.

The main benefit I can see is that JAX has few primitives with plaform-specific lowerings and I expect that a multi-platform lowering using conditionals for those primitives will be much smaller than multiple lowerings. The lowering time should similarly be smaller. Philosophically, such a lowering seems to fit better with JAX, that is mostly platform-independent and picks different lowerings only exceptionally to workaround differences in the XLA compilers. It is likely that there differences will get smaller.

However, I am not 100% certain that this scheme will work in all cases, it is somewhat of an experiment.

In terms of the costs of this scheme, there is a bit of extra complexity, but not much. I have already tested that custom calls are not an issue, because the HLO lowering already eliminates the redundant branches, and the resulting HLO that gets to the compiler is the same as if we lowering just for one platform.

The alternative, which I think you have in mind, is to add multiple serialized modules to the XlaCallModule op?

**Comment by maxwillzq** [comment]
Good discussion.  There is another user case maybe worth to mention here. In PAX/SAX, I saw some jax function has backend-conditional logic.  For example
```python
if jax.default_backend() == 'tpu':
  # do something
else:
 # do something else
```

I wonder how to do multi-platform lowering here now.   cc to @rxsang  and @voutcn.

**Comment by gnecula** [comment]
> Good discussion. There is another user case maybe worth to mention here. In PAX/SAX, I saw some jax function has backend-conditional logic. For example
> 
> ```python
> if jax.default_backend() == 'tpu':
>   # do something
> else:
>  # do something else
> ```
> 
> I wonder how to do multi-platform lowering here now. cc to @rxsang and @voutcn.

That code cannot be handled by multi-platform lowering, or cross-platform lowering. It is in fact very suspicious because it uses global device state to decide what code to use, but JAX will make platform decisions based on what arguments are passed to the function being traced.

We should look into why they are doing this and what are cleaner ways of achieving that. But this should be a separate issue.

**Comment by maxwillzq** [comment]
> 

I paste one example here  third_party/py/praxis/py_utils.py;l=457-466
